### DISCUSSION TITLE
streams_generator_i2s_ESP32 generates incorrect bit rate but MCLK is correct. How?

### DISCUSSION THREAD

**Discussion by groeneba** [pschatzmann/arduino-audio-tools#1453] [Q&A] [answered]
Would like to use ESP32 Dev Module as BT sink, which did not work, leading to this trial, which showed the problem.

Setup of test sinewave stream is:

```
AudioInfo info(48000, 2, 24);
SineWaveGenerator<int16_t> sineWave(16000);                // subclass of SoundGenerator with max amplitude of 32000
```

Set Audiologger to Warning:

` AudioLogger::instance().begin(Serial, AudioLogger::Warning); // BG change to Warning else sine is interrupted`

I added MCK signal in AudioConfig.h line 229:
`#define PIN_I2S_MCK 3           // BG edit `

I get 12.29MHz as MCK, but BCK and WCK is incorrect at 2.458MHz and 51.20 kHz respectively. Expected is 2.304MHz and 48 kHz.

This does work with PCM5102, generating a tone, BUT Texas Instruments CC8531 RF SoC refuses to accept this as valid input, because the sample rate is wrong. See attached scope screenshot with measured clock rates.

![image](https://github.com/pschatzmann/arduino-audio-tools/assets/26008067/c54c0ac6-3800-4cb1-be50-a98deb394064)

How do I fix this?

**Comment by pschatzmann** [comment, marked as answer]
I don't understand what you try to do and what your issue is: BlueTooth Sink and Sine Generator and a sample rate of 48000. This does not make any sense to me!

If you think that the frequency is wrong, then I think this would rather be a question for Espressif: The only thing that I can contribute is that you I2S lets you define a fixed_mclk: see https://docs.espressif.com/projects/esp-idf/en/v4.2.2/esp32/api-reference/peripherals/i2s.html

In addition your code is wrong: you generate a 16 bit signal and pretend that it is 24 bits. This does not generate any meaningful output: use 16 bits with int16_t and 24 bits with int24_t: 
```
AudioInfo info(48000, 2, 24);
SineWaveGenerator<int24_t> sineWave(16000);                // subclass of SoundGenerator with max amplitude of 32000
```

**Reply by groeneba** [reply]
Thanks for the reply. Less information then: I want to generate a signal on I2S bus using your "streams_generator_i2s_ESP32" Arduino sketch and will set the sine generator up as suggested. It should solve the issue. Appreciate your work!

**Comment by pschatzmann** [comment]
No worries: I got confused first, because initially I thought you wanted to use my A2DP library to implement the A2DP sink. This uses the IDF API that only supports 44100. I needed to read the text  a second time to finally understand what your issue is: sometimes less is more!

**Reply by groeneba** [reply]
Thanks again, we really appreciate the time you put into this. Now for the next silly question, reverting to the Bluetooth A2DP Sink. It works very well with 16bit and 32 bits per sample, but does not work with 24 bits per sample - no serial data output. I have not investigated why yet, but is it a limitation on the library or my understanding of I2S?

I used it with a PCM5102 DAC.
```

                                                                                          /*
  Streaming Music from Bluetooth
  
  Copyright (C) 2020 Phil Schatzmann
  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.
  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

// ==> Example A2DP Receiver which uses I2S to an external DAC

#include "BluetoothA2DPSink.h"

BluetoothA2DPSink a2dp_sink;

void setup() {

    static i2s_config_t i2s_config = {
      .mode = (i2s_mode_t) (I2S_MODE_MASTER | I2S_MODE_TX),
      .sample_rate = 41400, // updated automatically by A2DP
      .bits_per_sample = (i2s_bits_per_sample_t)24,
      .channel_format = I2S_CHANNEL_FMT_RIGHT_LEFT,
      .communication_format = (i2s_comm_format_t) (I2S_COMM_FORMAT_STAND_I2S),
      .intr_alloc_flags = 0, // default interrupt priority
      .dma_buf_count = 8,
      .dma_buf_len = 64,
      .use_apll = true,
      .tx_desc_auto_clear = true // avoiding noise in case of data unavailability
  };
  a2dp_sink.set_i2s_config(i2s_config);
  a2dp_sink.start("MyMusic");  

}


void loop() {
  delay(1000); // do nothing
}

```

**Comment by pschatzmann** [comment]
I would suggest to use the [AudioTools to convert the 16 bits to 24](https://github.com/pschatzmann/arduino-audio-tools/wiki/Converting-the-Data-Format). But to be honest, I don't see any advantage of going over 16...

**Reply by groeneba** [reply]
The 24 bit fixation is academic to be honest, because that is what the CC85XX can achieve, but thanks! Will play further. Enjoying your libraries!

**Reply by pschatzmann** [reply]
In the A2DP library there is a hack that if you define bits_per_sample other then 16, I am using [i2s_write_expand](https://docs.espressif.com/projects/esp-idf/en/v3.3/api-reference/peripherals/i2s.html#_CPPv416i2s_write_expand10i2s_port_tPKv6size_t6size_t6size_tP6size_t10TickType_t). 

But in any case the original source will always be 16 bits: Your just using the microcontroller to do the expansion instead of the DAC.
